text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,279 @@
+- **attachment_timestamp** the timestamp for when Proof-of-Work is completed = 9 trytes
+- **attachment_timestamp_lowerbound** is a slot for future use = 9 trytes
+- **attachment_timestamp_upperbound** is a slot for future use = 9 trytes
+- **tag** arbitrary user-defined value = 27 trytes
+- **obsolete_tag** another arbitrary user-defined tag = 27 trytes
+- **nonce** is the Proof-of-Work nonce of the transaction = 27 trytes
+- **bundle hash** is the hash of the entire bundle = 81 trytes
+- **current_index** the position of the transaction in its bundle = 9 trytes
+- **last_index** the total number of transactions in the bundle = 9 trytes
+
+A bundle is a collection of specific transactions. Bundles are required to bundle related information.
+For example, if not all data fits into one transaction, it has to be split across multiple transactions. An example would be payment settlements - they require the help of multiple transactions.
+A bundle then represents the collection of these transactions. It should be noted, that each transaction of a bundle will be sent separately through the network.
+
+Each transaction is identified by its hash, the transaction hash. Transactions should be final and therefore not modifiable after construction.
+The same applies to bundles. Bundles are final, and therefore not modifiable after construction.
+Having bundles as well as transactions final helps achieve correctness among the data objects.
+
+Transactions can be built from Transaction Builders.
+Bundles can be built from Bundle Builders. Both builder objects can be manipulated as desired.
+It should be noted, Bundle Builders don't process Transactions (as transactions are final), instead, they process Transaction Builders.
+
+## Exposed Interface
+
+### Transaction
+
+ A transaction consists of following fields:
+
+```rust
+struct Transaction {
+
+    transaction_hash: String,
+    signature_fragments: String,
+    address: String,
+    value: i64,
+    obsolete_tag: String,
+    timestamp: i64,
+    current_index: usize,
+    last_index: usize,
+    bundle_hash: String,
+    trunk: String,
+    branch: String,
+    nonce: String,
+    tag: String,
+    attachment_timestamp: i64,
+    attachment_timestamp_lower_bound: i64,
+    attachment_timestamp_upper_bound: i64,
+
+}
+```
+
+As mentioned, Transactions are final. Transaction fields therefore should be only accessible by getter functions.
+Besides that, Transactions can be only built by Transaction Builders, or by the bytes of a received transaction.
+
+```rust
+impl Transaction {
+
+    pub fn from_transaction_builder(transaction_builder: &TransactionBuilder) -> Result<Transaction, TransactionBuilderValidationError> {
+                
+        match TransactionBuilderValidator::validate(&transaction_builder) {
+            Ok(()) => { ; }
+            Err(e) => { return Err(e) }
+        }
+
+        let (transaction_hash, nonce, timestamp) = Pow::compute(&transaction_builder);
```

Comment:
IMHO this should be designed in a way that `Pow::compute(...)` returns a tuple, in this case a pair `(transaction, metadata)` where metadata contains the `transaction_hash` and the `timestamp`.

Reply:
(transaction, metadata), where metadata contains the `transaction_hash`, the `timestamp` from the last iteration directly before the `nonce` was found, and the found `nonce` itself.

Reply:
Btw, if we do the creation of `Transaction` in Pow, we need to declare the Transaction fields as public. Else we end in something like this [https://doc.rust-lang.org/error-index.html#E0451](https://doc.rust-lang.org/error-index.html#E0451)

Reply:
@rufsam yea, your're right. I don't know why I wrote the timestamp would be metadata. I mean it is kind of, right?, but as it's also part of the transaction itself, of course it would be part of it.

Reply:
I'm not sure if we need an additional timestamp in the metadata. I think we are fine with having it only in the transaction.